espnet2/train/gan_trainer.py

```diff
@@ -0,0 +1,470 @@
+#  Apache 2.0  (http://www.apache.org/licenses/LICENSE-2.0)
+
+"""Trainer module for GAN-based training."""
+
+import argparse
+import dataclasses
+import logging
+import time
+
+from contextlib import contextmanager
+from distutils.version import LooseVersion
+from pathlib import Path
+from typing import Dict
+from typing import Iterable
+from typing import List
+from typing import Optional
+from typing import Sequence
+from typing import Tuple
+from typing import Union
+
+import numpy as np
+import torch
+import wandb
+
+from torch.utils.tensorboard import SummaryWriter
+from typeguard import check_argument_types
+
+from espnet2.main_funcs.calculate_all_attentions import calculate_all_attentions
+from espnet2.schedulers.abs_scheduler import AbsBatchStepScheduler
+from espnet2.schedulers.abs_scheduler import AbsScheduler
+from espnet2.torch_utils.device_funcs import to_device
+from espnet2.torch_utils.recursive_op import recursive_average
+from espnet2.train.distributed_utils import DistributedOption
+from espnet2.train.reporter import SubReporter
+from espnet2.train.trainer import Trainer
+from espnet2.utils.build_dataclass import build_dataclass
+from espnet2.utils.types import str2bool
+
+if torch.distributed.is_available():
+    from torch.distributed import ReduceOp
+
+if LooseVersion(torch.__version__) >= LooseVersion("1.6.0"):
+    from torch.cuda.amp import autocast
+    from torch.cuda.amp import GradScaler
+else:
+    # Nothing to do if torch<1.6.0
+    @contextmanager
+    def autocast(enabled=True):
+        yield
+
+    GradScaler = None
+
+try:
+    import fairscale
+except ImportError:
+    fairscale = None
+
+
+@dataclasses.dataclass
+class TrainerOptions:
+    ngpu: int
+    resume: bool
+    use_amp: bool
+    train_dtype: str
+    grad_noise: bool
+    accum_grad: int
+    grad_clip: float
+    grad_clip_type: float
+    log_interval: Optional[int]
+    no_forward_run: bool
+    use_tensorboard: bool
+    use_wandb: bool
+    output_dir: Union[Path, str]
+    max_epoch: int
+    seed: int
+    sharded_ddp: bool
+    patience: Optional[int]
+    keep_nbest_models: Union[int, List[int]]
+    early_stopping_criterion: Sequence[str]
+    best_model_criterion: Sequence[Sequence[str]]
+    val_scheduler_criterion: Sequence[str]
+    unused_parameters: bool
+    wandb_model_log_interval: int
+    generator_first: bool
+
+
+class GANTrainer(Trainer):
+    """Trainer for GAN-based training.
+
+    If you'd like to use this trainer, the model must inherit
+    espnet.train.abs_gan_espnet_model.AbsGANESPnetModel.
+
+    """
+
+    @classmethod
+    def build_options(cls, args: argparse.Namespace) -> TrainerOptions:
+        """Build options consumed by train(), eval(), and plot_attention()"""
+        assert check_argument_types()
+        return build_dataclass(TrainerOptions, args)
+
+    @classmethod
+    def add_arguments(cls, parser: argparse.ArgumentParser):
+        parser.add_argument(
+            "--generator_first",
```

Comment:
Different way: `roles = ["generator", "discriminator"]`.
I know this is redundant but it can handle more than two models. I’ll leave it up to you.

Reply:
Thank you for your suggestion.
This is nice but the current abstract model assumes that two way forward.
So for now I will keep the current style.